casper/src/main/scala/coop/rchain/casper/util/comm/ApproveBlockProtocol.scala

```diff
@@ -0,0 +1,121 @@
+    val validSig = for {
+      c   <- a.candidate if c == this.candidate
+      sig <- a.sig if Validate.signature(sigData, sig)
+    } yield sig
+    val sender = a.sig.fold("")(sig => Base16.encode(sig.publicKey.toByteArray))
+
+    validSig.fold(Log[F].warn(s"APPROVAL: ignoring invalid block approval from $sender")) { sig =>
+      sigsF.modify(_ + sig) *> Log[F].info(s"APPROVAL: received block approval from $sender")
+    }
+  }
+
+  def currentSigs: F[Set[Signature]] = sigsF.get
+
+  def approvedBlock: F[Option[ApprovedBlock]] =
+    sigsF.get.map(
+      sigs =>
+        if (sigs.size >= candidate.requiredSigs) Some(ApprovedBlock(Some(candidate), sigs.toSeq))
+        else none[ApprovedBlock])
+
+  //TODO: potential optimization, only send to peers we have not
+  //      reveived a valid signature from yet
+  def sendUnapprovedBlock: F[Unit] =
+    for {
+      _ <- Log[F].info(s"APPROVAL: Beginning send of UnapprovedBlock $candidateHash to peers...")
+      _ <- CommUtil.sendToPeers[F](transport.UnapprovedBlock, serializedUnapprovedBlock)
+      _ <- Log[F].info(s"APPROVAL: Sent UnapprovedBlock $candidateHash to peers.")
+    } yield ()
+
+  def sendApprovedBlock(a: ApprovedBlock): F[Unit] = {
+    val serializedApprovedBlock = a.toByteString
+    val hash                    = a.candidate.flatMap(_.block).fold("")(b => PrettyPrinter.buildString(b.blockHash))
+    for {
+      _ <- Log[F].info(s"APPROVAL: Beginning send of ApprovedBlock $hash to peers...")
+      _ <- CommUtil.sendToPeers[F](transport.ApprovedBlock, serializedApprovedBlock)
+      _ <- Log[F].info(s"APPROVAL: Sent ApprovedBlock $hash to peers.")
+    } yield ()
+  }
+
+}
+
+object ApproveBlockProtocol {
+
+  def apply[F[_]](implicit instance: ApproveBlockProtocol[F]): ApproveBlockProtocol[F] = instance
+
+  def create[F[_]: Capture: Monad: NodeDiscovery: TransportLayer: Log: Time: ErrorHandler](
+      candidate: ApprovedBlockCandidate,
+      duration: Long): F[ApproveBlockProtocol[F]] =
+    Time[F].currentMillis.map(now => new ApproveBlockProtocol(candidate, now, duration))
+
+  def blockApprovalPacketHandler[F[_]: Monad: ApproveBlockProtocol](
+      peer: PeerNode): PartialFunction[Packet, F[Option[Packet]]] =
+    Function
+      .unlift(packetToBlockApproval)
+      .andThen {
+        case ba: BlockApproval =>
+          for {
+            _ <- ApproveBlockProtocol[F].addApproval(ba)
+          } yield none[Packet]
+      }
+
+  def run[F[_]: Monad: Capture: Time: ApproveBlockProtocol](waitTime: Long = 5000L): F[Unit] =
+    for {
+      t             <- Time[F].currentMillis
+      maybeApproved <- ApproveBlockProtocol[F].approvedBlock
+      _ <- maybeApproved match {
+            case Some(approvedBlock)
+                if (ApproveBlockProtocol[F].candidate.requiredSigs == 0 || t >= ApproveBlockProtocol[
+                  F].start + ApproveBlockProtocol[F].duration) =>
+              ApproveBlockProtocol[F].sendApprovedBlock(approvedBlock)
```

Comment:
Can you explain the reasoning behind `t >= ApproveBlockProtocol[F].start + ApproveBlockProtocol[F].duration`?

Reply:
We want to wait for the duration to elapse before creating the approved block. This gives an opportunity to collect signatures in excess of the minimum. Say for example there are 10 validators and we only need signatures from 8 to make an approved block. If we get the first 8 signatures within 5 seconds, we could make an approved block right then, but suppose waiting 5 more seconds would give us the remaining 2. Setting a duration of 10-15 seconds would give us all 10 signatures and still not be a terribly long time to wait for the approved block to be created.

Reply:
So if it hasn't passed that duration you want to call `              ApproveBlockProtocol[F].sendUnapprovedBlock *> IOUtil.sleep(waitTime) *> run[F](
                waitTime)` in an else case like below?

Reply:
That is exactly what gets called. `case _` will match both `None` and `Some(_)` when the `if t >= ...` returns false.